tools/federation/src/main/java/org/eclipse/rdf4j/federated/FedXConnection.java

```diff
@@ -138,10 +138,8 @@ public FedXConnection(FedX federation, FederationContext federationContext)
 
 		try {
 			// make sure to apply any external bindings
-			// Note: we use a conditional check for performance reasons.
-			// Further note that we subtract -1 from the size as the baseUri is optional
 			BindingSet queryBindings = EmptyBindingSet.getInstance();
-			if (bindings.size() > FedXRepositoryConnection.FEDX_BINDINGS.size() - 1) {
+			if (!FedXRepositoryConnection.FEDX_BINDINGS.containsAll(bindings.getBindingNames())) {
```

Comment:
Fwiw I believe this is still a check worth doing as in most cases it will skip creating a new MapBindingSet, and instead just pass in the `EmptyBindingSet` singleton - which other parts of the engine have some optimizations for.

Reply:
Good solution and way more robust. Should do the job as expected